app/ide-desktop/lib/client/src/authentication.ts

```diff
@@ -0,0 +1,140 @@
+ *
+ * To handle these redirects, we use deep links. Deep links are URLs that are used to redirect the
+ * user to a specific page in the application. To handle deep links, we use a custom URL protocol
+ * scheme.
+ *
+ * To prepare the application to handle deep links:
+ * - Register a custom URL protocol scheme with the OS (c.f., `electron-builder-config.ts`).
+ * - Define a listener for Electron {@link OPEN_URL_EVENT}s (c.f., {@link initOpenUrlListener}).
+ * - Define a listener for {@link ipc.channel.openDeepLink} events (c.f., `preload.ts`).
+ *
+ * Then when the user clicks on a deep link from an external source to the IDE:
+ * - The OS redirects the user to the application.
+ * - The application emits an Electron {@link OPEN_URL_EVENT}.
+ * - The {@link OPEN_URL_EVENT} listener checks if the {@link URL} is a deep link.
+ * - If the {@link URL} is a deep link, the {@link OPEN_URL_EVENT} listener prevents Electron from
+ * handling the event.
+ * - The {@link OPEN_URL_EVENT} listener then emits an {@link ipc.channel.openDeepLink} event.
+ * - The {@link ipc.channel.openDeepLink} listener registered by the dashboard receives the event.
+ * Then it parses the {@link URL} from the event's {@link URL} argument. Then it uses the
+ * {@link URL} to redirect the user to the dashboard, to the page specified in the {@link URL}'s
+ * `pathname`. */
+import * as electron from 'electron'
+import * as ipc from 'ipc'
+import * as shared from '../shared'
+import opener from 'opener'
+
+
+
+// =================
+// === Constants ===
+// =================
+
+/** Name of the Electron event that is emitted when a URL is opened in Electron (e.g., when the user
+ * clicks a link in the dashboard). */
+const OPEN_URL_EVENT = 'open-url'
+
+
+
+// ========================================
+// === Initialize Authentication Module ===
+// ========================================
+
+/** Configures all the functionality that must be set up in the Electron app to support
+ * authentication-related flows. Must be called in the Electron app `whenReady` event. */
+export const initAuthenticationModule = (
+    window: () => electron.BrowserWindow | null,
+) => {
+    initIpc()
+    initOpenUrlListener(window)
+}
+
+/** Registers an Inter-Process Communication (IPC) channel between the Electron application and the
+ * served website.
+ *
+ * This channel listens for {@link ipc.channel.openUrlInSystemBrowser} events. When this kind of
+ * event is fired, this listener will assume that the first and only argument of the event is a URL.
+ * This listener will then attempt to open the URL in a cross-platform way. The intent is to open
+ * the URL in the system browser.
+ *
+ * This functionality is necessary because we don't want to run the OAuth flow in the app. Users
+ * don't trust Electron apps to handle their credentials. */
+const initIpc = () => {
+    electron.ipcMain.on(ipc.channel.openUrlInSystemBrowser, (_event, url) => opener(url))
```

Comment:
So what `opener` does is it opens a browser with requested url in it. The library is a slim wrapper that detects on which platform is being used and calls right command:
- Windows: `cmd.exe`
- Mac: open
- Linux: xdg-open

In fact it's just couple lines of code: https://github.com/domenic/opener/blob/master/lib/opener.js

I also don't like installing a lib for single purpose only but in this case this is a useful snippet of code I think.

If you really insist I would vote for creating FIXME with a chore ticket not to block the whole PR

Reply:
I responded to that comment in the previous PR review. This library is ~60 LOC with 0 dependencies and it's used to make the opening of URLs work cross-platform. Yes, we could use built-in methods, but it'd be easier to vendor the dependency into the repo since we'd end up writing the same code. Is that OK?

Reply:
Oh, so this is actually doing some interesting logic there, like it's platform-specific. Ok, in such a case my comment is invalid. Thanks for pointing it out!